tree/tree/src/InternalTreeUtils.cxx

```diff
@@ -0,0 +1,221 @@
+/// \param[in] tree The tree from which friends information will be gathered.
+/// \throws std::runtime_error If no files could be associated with the input tree.
+std::vector<std::string> GetFileNamesFromTree(const TTree &tree)
+{
+   std::vector<std::string> filenames;
+
+   // If the input tree is a TChain, traverse its list of associated files.
+   if (auto chain = dynamic_cast<const TChain *>(&tree)) {
+      const auto *chainFiles = chain->GetListOfFiles();
+      if (!chainFiles) {
+         throw std::runtime_error("Could not retrieve a list of files from the input TChain.");
+      }
+      // Store this in a variable so it can be later used in `filenames.reserve`
+      // if it passes the check.
+      const auto nfiles = chainFiles->GetEntries();
+      if (nfiles == 0) {
+         throw std::runtime_error("The list of files associated with the input TChain is empty.");
+      }
+      filenames.reserve(nfiles);
+      for (const auto *f : *chainFiles)
+         filenames.emplace_back(f->GetTitle());
+   } else {
+      const TFile *f = tree.GetCurrentFile();
+      if (!f) {
+         throw std::runtime_error("The input TTree is not linked to any file, "
+                                  "in-memory-only trees are not supported.");
+      }
+
+      filenames.emplace_back(f->GetName());
+   }
+
+   return filenames;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// \fn RFriendInfo GetFriendInfo(const TTree &tree)
+/// \ingroup tree
+/// \brief Get and store the names, aliases and file names of the direct friends of the tree.
+/// \param[in] tree The tree from which friends information will be gathered.
+/// \throws std::runtime_error If the input tree has a list of friends, but any
+///         of them could not be associated with any file.
+///
+/// Calls TTree::GetListOfFriends and parses its result for the names, aliases
+/// and file names, with different methodologies depending on whether the
+/// parameter is a TTree or a TChain.
+///
+/// \note This function only retrieves information about <b>direct friends</b>
+///       of the input tree. It will not recurse through friends of friends and
+///       does not take into account circular references in the list of friends
+///       of the input tree.
+///
+/// \warning If any friend of the input tree is a TChain, this function expects
+/// that the chain has one single tree inside, i.e. the same tree name
+/// for all the files associated with the chain. Thus, a usecase like:
+/// ~~~{.cpp}
+/// TTree parent{"parent","parent"};
+/// TChain friendchain{};
+/// friendchain.Add("friendfile1.root/friendtree1");
+/// friendchain.Add("friendfile2.root/friendtree2");
+/// parent.AddFriend(&friendchain);
+/// auto friendinfo = GetFriendInfo(parent);
+/// ~~~
+/// \warning will just retrieve the name "friendtree1" but not "friendtree2" for
+/// the friend chain. A possible workaround is to define the friend
```

Comment:
Is that really a work-around?  I don't see it.  The first one put the 2 tree one after the other "vertically" (increasing the number of entries).  The second put the 2 tree as friends, i.e. horizontally, increasing the effective number of branches.

Reply:
I'm not sure I understand the point of having a friend TChain with different tree names then. I thought so far it was all about attaching data horizontally, i.e. this table like I posted in a comment above
| treeparent.x | treefriend1.x | treefriend2.x | treefriend3.x | treefriend4.x | treefriend5.x |
|--------------|---------------|---------------|---------------|---------------|---------------|
| 1            | 1             | 1             | 1             | 1             | 1             |
| 2            | 2             | 2             | 2             | 2             | 2             |
| 3            | 3             | 3             | 3             | 3             | 3             |
| 4            | 4             | 4             | 4             | 4             | 4             |
| 5            | 5             | 5             | 5             | 5             | 5             |

Which to me, and I admit I may be wrong, looks like it can be the result both of multiple trees attached as separate friends to the parent one, or of one single TChain made internally of different trees and that chain is attached as a single friend to the parent tree. See my gist above for code reference.
But from the data perspective I thought we were always talking about horizontally concatenating columns of a dataset.

If on the other hand we're talking about vertically aligning data that semantically belongs to the same column, I really find it difficult to justify having different dataset names (i.e. treenames) for different files, since it's always the same column we're dealing with, e.g.
| filename         | dataset   | column_x |
|------------------|-----------|----------|
| parentfile.root  | mydataset | 1        |
|                  |           | 2        |
|                  |           | 3        |
| friendfile1.root | mydataset | 4        |
|                  |           | 5        |
|                  |           | 6        |
| friendfile2.root | mydataset | 7        |
|                  |           | 8        |
|                  |           | 9        |

Reply:
And in my opinion this also raises the question of what's the meaning of `treefriend.x` if in theory  `treefriend.x` should be a vertical-continuation of `treeparent.x`, that is both columns should be named `x` and there should be no separation between them after the call of `parenttree->AddFriend(&friendtree)`

Reply:
Your last comment Vincenzo is exactly what I did not understand about the first version of your gist, I am not sure how RDF is supposed to deal with that

Reply:
> I'm not sure I understand the point of having a friend TChain with different tree names then. 

This maybe a case where the producer of the file changed the name of the tree in the middle of production (for example to fix a typo) but does not want (or can not) reproduce the old data files with the new names.  (or 2 colleagues created the semantically same TTree but use different name (one in French one in Italian :) ) and they want to combine their results.

Reply:
> looks like it can be the result both of multiple trees attached as separate friends to the parent one, or of one single TChain made internally of different trees and that chain is attached as a single friend to the parent tree. 

As Enrico said, Friends are for horizontal merge (more branches) and Chains are for vertical merge (same branches, more entries).

Reply:
> As Enrico said, Friends are for horizontal merge (more branches) and Chains are for vertical merge (same branches, more entries).
Sure I understand that, but what about a Friend TChain ? i.e. something similar to what is produced in my gist
```
TFile parentfile{"parentfile.root","read"};
std::unique_ptr<TTree> parent{parentfile.Get<TTree>("treeparent")};

TChain friendchain{};
friendchain.Add("friendfile1.root/friendtree1");
friendchain.Add("friendfile2.root/friendtree2");
friendchain.Add("friendfile3.root/friendtree3");

parent->AddFriend(&friendchain);
```
Seen from the perspective of the dataset, this leads to something like
| filename        | dataset    | parent_column_x | filename         | dataset     | friends_column_x |
|-----------------|------------|----------|------------------|-------------|----------|
| parentfile.root | treeparent | 1        | friendfile1.root | friendtree1 | 1        |
|                 |            | 2        |                  |             | 2        |
|                 |            | 3        |                  |             | 3        |
|                 |            |          | friendfile2.root | friendtree2 | 4        |
|                 |            |          |                  |             | 5        |
|                 |            |          |                  |             | 6        |
|                 |            |          | friendfile3.root | friendtree3 | 7        |
|                 |            |          |                  |             | 8        |
|                 |            |          |                  |             | 9        |

If this is correct, than I have these doubts:
1.  With TTree::Draw I am able to write "friendtree1.x" or "friendtree2.x", but what if I want to access the whole friendchain "x" column, i.e. what is called in the above table "friends_column_x" ? What should I write in code ?
2. TTree::Draw is able to plot `("x:friendtree1.x")`, so it means that internally it remembers that portions of the same column came from different "datasets", even though semantically they belong to the same column. I admit this was quite unexpected
3. I now appreciate more the RDataFrame error which I previously  didn't understand enough and I think it's correct.

Reply:
I now also understand better this comment by Enrico
> If so: friends with a different number of entries than the main trees are not supported, and I am not sure in general there is a good branch name that can be used for all branches called "x" in all friends, if the friend trees have different names and the TChain does not have a name.

which leads me to believe the answer to my doubt 1. above is that "friends_column_x" can be referenced by "FriendChainName.x" in code, but only if the user provided explicitly a name for the chain. What happens otherwise?

Reply:
> What happens otherwise?

Then the name is ambiguous and the branch can not be found by "name" lookup alone (but could be found by iterating through the list of friends and list of branches).